operator/pkg/reconciliation/decommission_node.go

```diff
@@ -0,0 +1,327 @@
+	// the endpoint data lingers for a while after it has been decommissioned
+	// so this scenario should be unlikely
+	return true
+}
+
+func (rc *ReconciliationContext) DeletePodPvcs(pod *v1.Pod) error {
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvcName := v.PersistentVolumeClaim.ClaimName
+		name := types.NamespacedName{
+			Name:      v.PersistentVolumeClaim.ClaimName,
+			Namespace: rc.Datacenter.Namespace,
+		}
+
+		podPvc := &corev1.PersistentVolumeClaim{}
+		err := rc.Client.Get(rc.Ctx, name, podPvc)
+		if err != nil {
+			rc.ReqLogger.Error(err, "Failed to get pod PVC", "Claim Name", pvcName)
+			return err
+		}
+
+		err = rc.Client.Delete(rc.Ctx, podPvc)
+		if err != nil {
+			rc.ReqLogger.Error(err, "Failed to delete pod PVC", "Claim Name", pvcName)
+			return err
+		}
+
+		rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.DeletedPvc,
+			"Claim Name: %s", pvcName)
+
+	}
+
+	return nil
+}
+
+func (rc *ReconciliationContext) RemoveDecommissionedPodFromSts(pod *v1.Pod) error {
+	podRack := pod.Labels[api.RackLabel]
+	var sts *appsv1.StatefulSet
+	for _, s := range rc.statefulSets {
+		if s.Labels[api.RackLabel] == podRack {
+			sts = s
+			break
+		}
+	}
+
+	if sts == nil {
+		// Failed to find the statefulset for this pod
+		return fmt.Errorf("Failed to find matching statefulSet for pod rack: %s", podRack)
+	}
+
+	maxReplicas := *sts.Spec.Replicas
+	lastPodSuffix := stsLastPodSuffix(maxReplicas)
+	if strings.HasSuffix(pod.Name, lastPodSuffix) {
+		return rc.UpdateRackNodeCount(sts, *sts.Spec.Replicas-1)
+	} else {
+		// Pod does not match the last pod in statefulSet
+		// This scenario should only happen if the pod
+		// has already been terminated
+		return nil
+	}
+}
+
+func stsLastPodSuffix(maxReplicas int32) string {
+	return fmt.Sprintf("sts-%v", maxReplicas-1)
+}
+
+func (rc *ReconciliationContext) EnsurePodsCanAbsorbDecommData(decommPod *v1.Pod, epData httphelper.CassMetadataEndpoints) error {
+	podsUsedStorage, err := rc.GetUsedStorageForPods(epData)
+	if err != nil {
+		return err
+	}
+
+	spaceUsedByDecommPod := podsUsedStorage[decommPod.Name]
+
+	for _, pod := range rc.dcPods {
+		if pod.Name == decommPod.Name {
+			continue
+		}
+
+		serverDataPv, err := rc.getServerDataPv(pod)
+		if err != nil {
+			return err
+		}
+
+		storageResource := serverDataPv.Spec.Capacity["storage"]
+		total := storageResource.AsDec().UnscaledBig().Int64()
+		used := podsUsedStorage[pod.Name]
+		free := total - int64(used)
+		if free < int64(spaceUsedByDecommPod) {
+			msg := "Not enough free space available to decommission"
+			rc.ReqLogger.Info(msg, "pod", pod.Name,
+				"free space", free,
+				"required space", spaceUsedByDecommPod)
+
+			err := fmt.Errorf(msg)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (rc *ReconciliationContext) GetUsedStorageForPods(epData httphelper.CassMetadataEndpoints) (map[string]float64, error) {
+	podStorageMap := make(map[string]float64)
```

Comment:
For my own enlightenment, what is the benefit of `make(map[string]float64)` over the literal `map[string]float64{}`?

Reply:
AFAIK there is no difference, other than the `{}` version lets you specify initial elements and the `make()` version lets you specify a capacity. In the case in which you are doing neither, I suppose it is just preference (in which case, I don't mind either way if anyone have a preference).

Reply:
Ah okay, I have no preference either, I was just curious.